File: script/make/protocol/js/FashionProtocol.js
export default class FashionProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 12001: {
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 12001: {
                // 时装列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Fashion
                    // 时装ID
                    const fashionId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 过期时间
                    const expireTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const fashion = {fashionId, expireTime};
                    // add
                    list.push(fashion);
                }
                return {list};
            }
            case 12002: {
                // 时装ID列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Fashion
                    // 时装ID
                    const fashionId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const fashion = {fashionId};
                    // add
                    list.push(fashion);
                }
                return {list};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/lua/FashionProtocol.lua
FashionProtocol = {}

function FashionProtocol.encode(offset, protocol, data)
    if protocol == 12001 then
        local offset = offset
        local table = {}
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function FashionProtocol.decode(offset, protocol, data)
    if protocol == 12001 then
        local offset = offset
        -- 时装列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Fashion
            -- 时装ID
            local fashionId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 过期时间
            local expireTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- object
            local fashion = {fashionId = fashionId, expireTime = expireTime}
            list[listIndex] = fashion
        end
        return {list = list}
    elseif protocol == 12002 then
        local offset = offset
        -- 时装ID列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- Fashion
            -- 时装ID
            local fashionId = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- object
            local fashion = {fashionId = fashionId}
            list[listIndex] = fashion
        end
        return {list = list}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end